packages/gatsby/src/internal-plugins/prod-404/gatsby-node.js

```diff
@@ -1,13 +1,28 @@
-let created404 = false
+const { emitter } = require(`../../redux`)
+const { boundActionCreators } = require(`../../redux/actions`)
+
+const PROD_404_PAGE_PATH = `/404.html`
+
+let page404 = null
 exports.onCreatePage = ({ page, store, actions }) => {
   // Copy /404/ to /404.html as many static site hosts expect
   // site 404 pages to be named this.
   // https://www.gatsbyjs.org/docs/add-404-page/
-  if (!created404 && /^\/?404\/?$/.test(page.path)) {
+  if (!page404 && /^\/?404\/?$/.test(page.path)) {
     actions.createPage({
       ...page,
-      path: `/404.html`,
+      path: PROD_404_PAGE_PATH,
     })
-    created404 = true
+    page404 = page
   }
 }
+
+emitter.on(`DELETE_PAGE`, action => {
+  if (page404 && action.payload.path === page404.path) {
+    boundActionCreators.deletePage({
+      ...page404,
+      path: PROD_404_PAGE_PATH,
+    })
+    page404 = null
+  }
+})
```

Comment:
- `prod-404` internal plugin is actually also used in `gatsby develop`
- removing `/404/` page (that this plugin clones into `/404.html`) wasn't removing clone which meant that 404 page removals didn't actually removed `/404.html` page

Reply:
I don't know much about this plugin, just want to understand: why does it hook into `onCreatePage` instead of creating its own page? And where do we actually create the source `404` page?

Reply:
Ok, so explanation for this plugin - when user creates `src/pages/404.js`, the actual path for page is `/404/` (that's what `gatsby-plugin-page-creator` does) which translated to `/404/index.html` when generating .html file - this is fine, but a lot of hosting solutions handle `/404.html` as potential default to show 404 page (I think Netlify does this for sure). So we do special case 404 kind of page and create "duplicate" page for it

The `src/pages/404.js` is also not only case when we do this - if user creates page programatically in `createPages` - this will work too.

Note that this is mostly needed for production builds, but our shared runtime (both dev and prod) also uses `/404.html` in multiple places, so the way this plugin works impact both develop and prod.